### PR TITLE
WIP: Client rotate remote key

### PR DESCRIPTION
**Pull request by cyli**
This supercedes https://github.com/docker/notary/pull/529.

This adds the functionality for the client to rotate remote keys, including the timestamp key.

In order to do so, it must sign a request with the root key, in order to prove that it has the root key, and the rotation is published right away.

Part of addressing #347

### CODE REVIEW

File: client/backwards_compatibility_test.go [hunk 1]
```diff
@@ -107,12 +107,13 @@ func Test0Dot1RepoFormat(t *testing.T) {
 	// one and try to create a new one from scratch, which will be the wrong version
 	require.NoError(t, repo.fileStore.RemoveMeta(data.CanonicalTimestampRole))
 
-	// rotate the timestamp key, since the server doesn't have that one
+	// rotate the timestamp key, since the server doesn't have that one (we can't just
```
Comment (diogomonica) [inline client/backwards_compatibility_test.go]: Confusing comment.

File: client/client.go [hunk 2]
```diff
@@ -829,28 +839,65 @@ func (r *NotaryRepository) validateRoot(rootJSON []byte) (*data.SignedRoot, erro
...
 func (r *NotaryRepository) RotateKey(role string, serverManagesKey bool) error {
-	if role == data.CanonicalRootRole || role == data.CanonicalTimestampRole {
+	if role == data.CanonicalRootRole {
 		return fmt.Errorf(
 			"notary does not currently support rotating the %s key", role)
 	}
 	if serverManagesKey && role == data.CanonicalTargetsRole {
 		return ErrInvalidRemoteRole{Role: data.CanonicalTargetsRole}
 	}
+	if !serverManagesKey && role == data.CanonicalTimestampRole {
```
Comment (diogomonica) [inline client/client.go]: We should probably add a comment before these ifs describing that these are our current assumptions: Targets is always local and timestamps is always remote.

File: client/client.go [hunk 3]
```diff
@@ -829,28 +839,65 @@ func (r *NotaryRepository) validateRoot(rootJSON []byte) (*data.SignedRoot, erro
...
 	)
 	if serverManagesKey {
-		pubKey, err = getRemoteKey(r.baseURL, r.gun, role, r.roundTrip)
+		if err := r.bootstrapRepo(); err != nil {
+			return err
+		}
+
+		// get root keys - since we bootstrapped the repo, a valid root should always specify
+		// the root role
+		rootRole, ok := r.tufRepo.Root.Signed.Roles[data.CanonicalRootRole]
```
Comment (diogomonica) [inline client/client.go]: not sure if `rootRole` is a good name for a list of root keys.

Comment (diogomonica) [inline client/client.go]: What about having a `r.getRootKeys()`

Comment (cyli) [inline client/client.go]: I was thinking about adding a `GetSigningKeys` to `tuf.Repo`, which already has logic similar to this for getting signing keys for signing a particular role.

File: client/client.go [hunk 4]
```diff
@@ -829,28 +839,65 @@ func (r *NotaryRepository) validateRoot(rootJSON []byte) (*data.SignedRoot, erro
...
+		if !ok {
+			return fmt.Errorf("update resulted in an invalid state with an invalid root")
```
Comment (diogomonica) [inline client/client.go]: Let's turn this into a method that returns `err` with something more descriptive.

File: client/client.go [hunk 5]
```diff
@@ -829,28 +839,65 @@ func (r *NotaryRepository) validateRoot(rootJSON []byte) (*data.SignedRoot, erro
...
+		}
+		rootKeys := make([]data.PublicKey, 0, len(rootRole.KeyIDs))
```
Comment (diogomonica) [inline client/client.go]: This should all be replaced by `getRootKeys()`

File: client/client.go [hunk 6]
```diff
@@ -829,28 +839,65 @@ func (r *NotaryRepository) validateRoot(rootJSON []byte) (*data.SignedRoot, erro
...
 	var (
-		pubKey data.PublicKey
-		err    error
+		pubKey       data.PublicKey
+		errFmtString string
+		err          error
 	)
 	if serverManagesKey {
-		pubKey, err = getRemoteKey(r.baseURL, r.gun, role, r.roundTrip)
+		if err := r.bootstrapRepo(); err != nil {
```
Comment (diogomonica) [inline client/client.go]: why `:=`?

File: client/client.go [hunk 7]
```diff
@@ -829,28 +839,65 @@ func (r *NotaryRepository) validateRoot(rootJSON []byte) (*data.SignedRoot, erro
 // creates and adds one new key or delegates managing the key to the server.
 // These changes are staged in a changelist until publish is called.
 func (r *NotaryRepository) RotateKey(role string, serverManagesKey bool) error {
```
Comment (diogomonica) [inline client/client.go]: This method could probably use some `Debug` output.

File: client/client.go (hunk 6 above)
Comment (cyli) [inline client/client.go]: This error isn't propagated to the temporary variable anyway, since it gets returned immediately.  I'm happy to make it `=` though.

File: client/client.go [hunk 8]
```diff
@@ -829,28 +839,65 @@ func (r *NotaryRepository) validateRoot(rootJSON []byte) (*data.SignedRoot, erro
...
 	if err != nil {
+		return fmt.Errorf(errFmtString, err)
+	}
+
+	if err = r.rootFileKeyChange(role, changelist.ActionCreate, pubKey); err != nil {
 		return err
 	}
 
-	return r.rootFileKeyChange(role, changelist.ActionCreate, pubKey)
+	if serverManagesKey {
```
Comment (diogomonica) [inline client/client.go]: Comment to the effect: `Key rotation is an offline operation unless we are rotating a remote key, in which case we need to ensure the rotation actually happens`

File: cmd/notary/keys.go [hunk 9]
```diff
@@ -361,13 +361,11 @@ func (k *keyCommander) keysRotate(cmd *cobra.Command, args []string) error {
 		rolesToRotate = []string{data.CanonicalSnapshotRole}
 	case data.CanonicalTargetsRole:
 		rolesToRotate = []string{data.CanonicalTargetsRole}
+	case data.CanonicalTimestampRole:
```
Comment (diogomonica) [inline cmd/notary/keys.go]: Does it still make sense to keep the default behavior for `rotate` with no arguments? (targets and snapshot?)

Comment (cyli) [inline cmd/notary/keys.go]: I think only if we care about backwards compatibility - if we don't, I'm fine with key rotation to require a role.  (Also, does it make sense to allow the user to manage their own timestamp key?  They can really screw themselves over, but if they're managing everything...)

Comment (cyli) [inline cmd/notary/keys.go]: Although maybe changing how CLI key rotation works should be in a different PR?

File: server/handlers/default.go [hunk 10]
```diff
@@ -143,70 +143,95 @@ func DeleteHandler(ctx context.Context, w http.ResponseWriter, r *http.Request)
...
 	}
-
-	logger := ctxu.GetLoggerWithField(ctx, gun, "gun")
 
 	s := ctx.Value("metaStore")
 	store, ok := s.(storage.MetaStore)
 	if !ok || store == nil {
-		logger.Error("500 GET storage not configured")
-		return errors.ErrNoStorage.WithDetail(nil)
+		return nil, errors.ErrNoStorage.WithDetail(nil)
```
Comment (diogomonica) [inline server/handlers/default.go]: Is it useful here to return some detail in case a user gets the error and pastes it to us?

File: server/handlers/default.go [hunk 11]
```diff
@@ -143,70 +143,95 @@ func DeleteHandler(ctx context.Context, w http.ResponseWriter, r *http.Request)
...
 	}
+
 	c := ctx.Value("cryptoService")
 	crypto, ok := c.(signed.CryptoService)
 	if !ok || crypto == nil {
-		logger.Error("500 GET crypto service not configured")
-		return errors.ErrNoCryptoService.WithDetail(nil)
+		return nil, errors.ErrNoCryptoService.WithDetail(nil)
```
Comment (diogomonica) [inline server/handlers/default.go]: Same here. Even if its an opaque error it might be useful to distinguish cryptoservice errors in the wild (vs storage errors, for example)

File: client/client.go (hunk 8 above)
Comment (cyli) [inline client/client.go]: Done!

File: client/client.go (hunk 7 above)
Comment (cyli) [inline client/client.go]: I added one Debug, and otherwise just tried to return better errors.  Does it look reasonable atm?

File: cmd/notary/keys.go (hunk 9 above)
Comment (endophage) [inline cmd/notary/keys.go]: I feel like we should change it, but I'm ok with it being a secondary PR

File: client/client.go [hunk 12]
```diff
@@ -55,6 +55,17 @@ func (err ErrInvalidRemoteRole) Error() string {
...
 
+// ErrInvalidLocalRole is returned when the client wants to manage
+// an unsupported key type
+type ErrInvalidLocalRole struct {
+	Role string
+}
+
+func (err ErrInvalidLocalRole) Error() string {
+	return fmt.Sprintf(
+		"notary does not support the client managing the %s key", err.Role)
```
Comment (endophage) [inline client/client.go]: Should the messaging be stronger? "notary does not permit ..."? Don't want people filing issues for things we never intend to support.

File: client/client.go [hunk 13]
```diff
@@ -829,28 +839,68 @@ func (r *NotaryRepository) validateRoot(rootJSON []byte) (*data.SignedRoot, erro
 // creates and adds one new key or delegates managing the key to the server.
 // These changes are staged in a changelist until publish is called.
 func (r *NotaryRepository) RotateKey(role string, serverManagesKey bool) error {
-	if role == data.CanonicalRootRole || role == data.CanonicalTimestampRole {
+	if role == data.CanonicalRootRole {
 		return fmt.Errorf(
 			"notary does not currently support rotating the %s key", role)
 	}
+	// We currently support locally managing targets keys, remotely managing
+	// timestamp keys, and locally or remotely managing timestamp keys.
```
Comment (endophage) [inline client/client.go]: "locally or remotely managing _snapshot_ keys"

File: client/client.go [hunk 14]
```diff
@@ -829,28 +839,68 @@ func (r *NotaryRepository) validateRoot(rootJSON []byte) (*data.SignedRoot, erro
 // creates and adds one new key or delegates managing the key to the server.
 // These changes are staged in a changelist until publish is called.
 func (r *NotaryRepository) RotateKey(role string, serverManagesKey bool) error {
-	if role == data.CanonicalRootRole || role == data.CanonicalTimestampRole {
+	if role == data.CanonicalRootRole {
```
Comment (endophage) [inline client/client.go]: I'd be inclined to turn these into a `switch role`

File: tuf/store/offlinestore.go [hunk 15]
```diff
@@ -38,7 +41,12 @@ func (es OfflineStore) RemoveMeta(name string) error {
 }
 
 // GetKey returns ErrOffline
-func (es OfflineStore) GetKey(role string) ([]byte, error) {
+func (es OfflineStore) GetKey(role string) (data.PublicKey, error) {
+	return nil, err
```
Comment (endophage) [inline tuf/store/offlinestore.go]: Looking at this again, I shouldn't have use the name `err`. It's not breaking anything now but it's a package scope var. Do you mind updating it to `errOffline` or something else less generic?

File: tuf/store/interfaces.go [hunk 16]
```diff
@@ -12,7 +17,8 @@ type MetadataStore interface {
 
 // PublicKeyStore must be implemented by a key service
 type PublicKeyStore interface {
-	GetKey(role string) ([]byte, error)
+	GetKey(role string) (data.PublicKey, error)
+	RotateKey(role string, cs signed.CryptoService, roots ...data.PublicKey) (data.PublicKey, error)
```
Comment (endophage) [inline tuf/store/interfaces.go]: The stores are meant to be reasonably low level and it feels like they're losing cohesion when they start referencing `PublicKey` and `CryptoService`.

File: client/client.go [hunk 17]
```diff
@@ -829,28 +839,68 @@ func (r *NotaryRepository) validateRoot(rootJSON []byte) (*data.SignedRoot, erro
...
+		rootKeys, err := r.tufRepo.SigningKeysForRole(data.CanonicalRootRole)
+		if err != nil {
+			logrus.Debug("The root is invalid, because it does not contain any signing keys")
+			return err
+		}
+
+		errFmtString = "unable to rotate remote key: %s"
+		remote, err := getRemoteStore(r.baseURL, r.gun, r.roundTrip)
+		if err == nil {
+			pubKey, err = remote.RotateKey(role, r.CryptoService, rootKeys...)
```
Comment (endophage) [inline client/client.go]: I think the signing functionality currently contained in `remote.RotateKey` should be happening here and the stores should continue to receive and return on `[]byte`

File: const.go [hunk 18]
```diff
@@ -18,4 +18,7 @@ const (
 	Sha256HexSize = 64
 	// TrustedCertsDir is the directory, under the notary repo base directory, where trusted certs are stored
 	TrustedCertsDir = "trusted_certificates"
+
+	// HTTPStatusTooManyRequests is the http status 429 - will be an exported constant in http library in Go 1.6
+	HTTPStatusTooManyRequests = 429
```
Comment (endophage) [inline const.go]: :-/ I'm disappointed that net/http doesn't already have a const for this. Looks like the RFC (6585) still hasn't been accepted after almost 4 years.

File: server/errors/errors.go [hunk 19]
```diff
@@ -83,8 +84,14 @@ var (
...
 		Value:          "NO_KEYALGORITHM",
 		Message:        "The server does not have a key algorithm configured.",
-		Description:    "No key algorihtm has been configured for the server and it has been asked to perform an operation that requires generation.",
+		Description:    "No key algorithm has been configured for the server and it has been asked to perform an operation that requires generation.",
 		HTTPStatusCode: http.StatusInternalServerError,
 	})
+	ErrCannotRotateKey = errcode.Register(errGroup, errcode.ErrorDescriptor{
+		Value:          "CANNOT_ROTATE_KEY",
+		Message:        "Key has already been rotated recently.",
+		Description:    "The key has been rotated too recently, and cannot be rotated again at this time.",
```
Comment (endophage) [inline server/errors/errors.go]: This reads oddly to me. Maybe "Cannot rotate, last key rotation too recent."

File: server/errors/errors.go [hunk 20]
```diff
@@ -83,8 +84,14 @@ var (
 	ErrNoKeyAlgorithm = errcode.Register(errGroup, errcode.ErrorDescriptor{
 		Value:          "NO_KEYALGORITHM",
 		Message:        "The server does not have a key algorithm configured.",
-		Description:    "No key algorihtm has been configured for the server and it has been asked to perform an operation that requires generation.",
+		Description:    "No key algorithm has been configured for the server and it has been asked to perform an operation that requires generation.",
 		HTTPStatusCode: http.StatusInternalServerError,
 	})
+	ErrCannotRotateKey = errcode.Register(errGroup, errcode.ErrorDescriptor{
```
Comment (endophage) [inline server/errors/errors.go]: Given there are lots of reasons a key rotation can fail, should the name be more specific like `ErrKeyRotationRateLimited`?

File: server/handlers/default.go [hunk 21]
```diff
@@ -143,70 +143,95 @@ func DeleteHandler(ctx context.Context, w http.ResponseWriter, r *http.Request)
...
+	switch s.role { // parseKeyParams ensures it's only timestamp or snapshot
 	case data.CanonicalTimestampRole:
-		key, err = timestamp.GetOrCreateTimestampKey(gun, store, crypto, keyAlgorithm)
+		key, err = timestamp.GetOrCreateTimestampKey(s.gun, s.store, s.crypto, s.keyAlgo)
 	case data.CanonicalSnapshotRole:
-		key, err = snapshot.GetOrCreateSnapshotKey(gun, store, crypto, keyAlgorithm)
-	default:
-		logger.Errorf("400 GET %s key: %v", role, err)
-		return errors.ErrInvalidRole.WithDetail(role)
+		key, err = snapshot.GetOrCreateSnapshotKey(s.gun, s.store, s.crypto, s.keyAlgo)
```
Comment (endophage) [inline server/handlers/default.go]: The default case seems to have dropped off? Should be maintained. There's nothing stopped a client (or even a curl request) from attempting to rotate the root, targets, or delegation keys.

Comment (endophage) [inline server/handlers/default.go]: Ah, I see you moved it above to line 179. I'm not sure exactly what the implementation of server signing will look like but I feel like it's simpler to keep it here in the switch. Then manipulating which roles can be server side is better contained in this one place and adding/removing cases is sufficient to change which keys the server holds, and there isn't a conditional elsewhere to worry about.

File: server/handlers/default.go [hunk 22]
```diff
@@ -143,70 +143,95 @@ func DeleteHandler(ctx context.Context, w http.ResponseWriter, r *http.Request)
...
-		return errors.ErrNoCryptoService.WithDetail(nil)
+		return nil, errors.ErrNoCryptoService.WithDetail("crypto service not configured")
 	}
+
 	algo := ctx.Value("keyAlgorithm")
 	keyAlgo, ok := algo.(string)
-	if !ok || keyAlgo == "" {
-		logger.Error("500 GET key algorithm not configured")
-		return errors.ErrNoKeyAlgorithm.WithDetail(nil)
+	if !ok || keyAlgo != data.ECDSAKey && keyAlgo != data.RSAKey && keyAlgo != data.ED25519Key {
```
Comment (endophage) [inline server/handlers/default.go]: I'd just stick with the original check and let the signer tell us if the requested algo is unsupported. This is just adding another place we have to remember to change if we tweak the signer. Especially as at any given time our signer actually only supports one of the algos.

**Comment by endophage** [comment]
I don't see where the server is checking the client's signed request. Also seems like a partial server implementation that currently just returns the existing key (if there is one). That's fine if the server piece is coming separately but I'd like to see the signed request checking in this PR given that functionality is landing here.

File: client/client.go (hunk 17 above)
Comment (diogomonica) [inline client/client.go]: I'm of the opposite opinion :D

File: client/client.go [hunk 23]
```diff
@@ -829,28 +839,68 @@ func (r *NotaryRepository) validateRoot(rootJSON []byte) (*data.SignedRoot, erro
...
+		} else if _, ok := err.(signed.ErrNoKeys); ok {
+			err = fmt.Errorf("root signing key unavailable so unable to rotate key anyway")
```
Comment (diogomonica) [inline client/client.go]: remove "anyway".

File: tuf/store/httpstore.go [hunk 24]
```diff
@@ -270,13 +275,54 @@ func (s HTTPStore) buildURL(uri string) (*url.URL, error) {
...
+func (s HTTPStore) requestKey(role string, body io.Reader) (data.PublicKey, error) {
 	url, err := s.buildKeyURL(role)
 	if err != nil {
 		return nil, err
 	}
-	req, err := http.NewRequest("GET", url.String(), nil)
+
+	method := "GET"
+	resource := role + " key"
+	if body != nil {
```
Comment (diogomonica) [inline tuf/store/httpstore.go]: Not sure if we should have implicit behavior like this.

File: tuf/store/httpstore.go [hunk 25]
```diff
@@ -270,13 +275,54 @@ func (s HTTPStore) buildURL(uri string) (*url.URL, error) {
...
+	req := &data.SignedCommon{
+		Type:    role,
+		Expires: time.Now().Add(5 * time.Minute),
+	}
+	reqJSON, err := json.Marshal(req)
+	if err != nil {
+		return nil, err
+	}
+	signedReq := &data.Signed{Signed: reqJSON}
+	if err := signed.Sign(cs, signedReq, roots...); err != nil {
```
Comment (diogomonica) [inline tuf/store/httpstore.go]: @endophage signature happening here.

File: cmd/notary/keys.go (hunk 9 above)
Comment (cyli) [inline cmd/notary/keys.go]: Filed https://github.com/docker/notary/issues/554 for it.  Basically am trying to keep this PR under 1k lines :D

File: tuf/store/interfaces.go (hunk 16 above)
Comment (cyli) [inline tuf/store/interfaces.go]: I agree that I wasn't super happy the arguments for `RotateKey` - perhaps a `KeyRotationHTTPStore` (or just `HTTPStore`) can take the `CryptoService`, etc., and can generate a key rotation request based on that?  Then the `RotateKey` parameters would just be the role as a string.

How do you feel about these functions returning `data.PublicKey` though instead of `[]byte`?  For metatdata especially, it makes sense for a remote store interface to return bytes, since we have to calculate the hash of those bytes.   But since no validation is done on the key bytes other than to make sure they're unmarshallable as keys, it seems anything that wants to use this interface ultimately cares about the keys, not the bytes themselves.  So it seems to make sense for the interface to return `data.PublicKey`?

Comment (cyli) [inline tuf/store/interfaces.go]: NM, sorry, just reread https://github.com/docker/notary/pull/553#discussion_r52268867, which says they should return `[]byte` :)  

Hmm... If we were using GRPC, would you still say that `GetKey` and `RotateKey` should return `[]bytes` and not `data.PublicKey`?  For instance I'd think that `GetMeta` and `SetMeta` should still take and return `[]bytes`, since that's what TUF operates on (to do checksumming, etc.)

File: tuf/store/httpstore.go (hunk 25 above)
Comment (endophage) [inline tuf/store/httpstore.go]: yeah... this feels like completely the wrong place to do this. The stores were very cohesive before. Now suddenly they're responsible for signing and json marshalling? This is how we end up with confusing code where it's not clear what happens where.

File: server/handlers/default.go (hunk 22 above)
Comment (cyli) [inline server/handlers/default.go]: Good point

File: client/client.go (hunk 12 above)
Comment (riyazdf) [inline client/client.go]: I'd agree with @endophage for stronger messaging, since we similar language for the root rotation error message and we intend to support that

File: client/client_test.go [hunk 26]
```diff
@@ -21,6 +21,7 @@ import (
 	"github.com/Sirupsen/logrus"
 	ctxu "github.com/docker/distribution/context"
 	"github.com/docker/go/canonical/json"
+	"github.com/gorilla/mux"
```
Comment (riyazdf) [inline client/client_test.go]: thanks for updating this :+1:

File: client/client_test.go [hunk 27]
```diff
@@ -2567,6 +2570,52 @@ func testRotateKeySuccess(t *testing.T, serverManagesSnapshotInit bool,
...
+// entirely
+func TestRemoteRotationNonRateLimitError(t *testing.T) {
+	ts, _, _ := simpleTestServer(t)
+	defer ts.Close()
+
+	repo, _ := initializeRepo(t, data.ECDSAKey, "docker.com/notary", ts.URL, true)
+	defer os.RemoveAll(repo.baseDir)
+
+	// simpleTestServer has no rotate key endpoint, so this should fail
+	assert.Error(t, repo.RotateKey(data.CanonicalTimestampRole, true))
```
Comment (riyazdf) [inline client/client_test.go]: Can we assert the type of error here?

File: cmd/notary/keys_test.go [hunk 28]
```diff
@@ -227,11 +233,10 @@ func TestRemoveMultikeysRemoveOnlyChosenKey(t *testing.T) {
 	}
 }
 
-// Non-roles, root, and timestamp can't be rotated
+// Non-roles, and root can't be rotated
```
Comment (riyazdf) [inline cmd/notary/keys_test.go]: I think it'd be good to also document delegation key behavior here

File: server/errors/errors.go (hunk 19 above)
Comment (riyazdf) [inline server/errors/errors.go]: +1 for @endophage's suggestion

File: tuf/store/httpstore.go (hunk 25 above)
Comment (cyli) [inline tuf/store/httpstore.go]: I guess it depends what you think of the signed request as.  If it should be part of any rotate key API, no matter if we implement a GRPC remote API, an in-memory API, etc, that all requires a signed request, and (and maybe in fact clients should never be able to rotate any key, even locally, without having the root key on hand?), then I agree it should probably go into the client.

If we are thinking of this as just something our HTTP server requires, as in it's part of the server's API and not a generic `RemoteStore` API, then it should probably live in the client for the notary server, which seems to be `HTTPStore`, since that's the thing which knows how to speak the server API.

File: server/handlers/default.go (hunk 21 above)
Comment (cyli) [inline server/handlers/default.go]: I don't mind moving it back, but I was going to have a `rotateKeyHandler`, which would need to perform the same check, so if we want it in a single location, should it be in `parseKeys`?  `https://github.com/docker/notary/pull/534/files#diff-b60e09f129d59867e23c3c6e1e94319cR159` was eventually supposed to be how this would look.

I can also just move it back to `getKeyHandler` for now, and when/if that change lands, move it to `parseKeyParams`

File: tuf/store/interfaces.go (hunk 16 above)
Comment (endophage) [inline tuf/store/interfaces.go]: If we were using GRPC there would have to be translation between the wire types (GRPC would require us to define our own protos), and the TUF types. The manually created part of the server and signer code would do that translation and sits a level higher than the "store" code.

So to answer you first question, I would definitely be happier to see something wrapped around the HTTPStore that could do the signing/validation.

File: tuf/store/httpstore.go (hunk 25 above)
Comment (endophage) [inline tuf/store/httpstore.go]: Yeah, the other store methods take or return a complete body (ok, for SetMulti it takes a map of part name to body). This method has to actually construct a body from the parameters it's passed.

If the signing was in an HTTP header or something more generic that could be included on any request, I'd feel better about putting it into the store.

File: server/handlers/default.go (hunk 21 above)
Comment (endophage) [inline server/handlers/default.go]: Switches without a default make me look twice, and are more concerning when dealing with user input. Having the default here makes it obvious what the valid input is for this flow. Having the check in parseKeyParams leaves it open to dropping off if somebody isn't paying attention.

File: tuf/store/interfaces.go (hunk 16 above)
Comment (cyli) [inline tuf/store/interfaces.go]: Hm, so I can see HTTPStore being just `[]bytes`, but do you think the interface should also just be `[]bytes`?  There can be something else between `HTTPStore` and the client, and `HTTPStore` doesn't have to definitely implement the interface.  Alternately, maybe `HTTPStore` can just implement the metadata part and something else that wraps it can implement the signing/unmarshalling?

**Comment by cyli** [comment]
@endophage Yep I was hoping to do the server part in a separate PR, because it's huge :(  And this is already huge.  Will be happy to add the server validation of the request, though - hopefully that shouldn't add to much.

**Comment by diogomonica** [comment]
@cyli rebase needed.

**Comment by cyli** [comment]
@diogomonica will do - I think in order to decrease the size of this I'm going to address the CLI changes here and #554 first in a separate PR :)

**Comment by cyli** [comment]
Closing in favor of https://github.com/docker/notary/pull/889
